tuftool/src/main.rs

```diff
@@ -92,70 +91,104 @@ enum Command {
-            Command::Delegation(cmd) => cmd.run(),
-            Command::Clone(cmd) => cmd.run(),
-            Command::TransferMetadata(cmd) => cmd.run(),
+            Command::Create(args) => args.run().await,
+            Command::Root(root_subcommand) => root_subcommand.run().await,
+            Command::Download(args) => args.run().await,
+            Command::Update(args) => args.run().await,
+            Command::Delegation(cmd) => cmd.run().await,
+            Command::Clone(cmd) => cmd.run().await,
+            Command::TransferMetadata(cmd) => cmd.run().await,
         }
     }
 }
 
-fn load_file<T>(path: &Path) -> Result<T>
+async fn load_file<T>(path: &Path) -> Result<T>
 where
     for<'de> T: serde::Deserialize<'de>,
 {
-    serde_json::from_reader(File::open(path).context(error::FileOpenSnafu { path })?)
-        .context(error::FileParseJsonSnafu { path })
+    serde_json::from_slice(
+        &tokio::fs::read(path)
+            .await
+            .context(error::FileOpenSnafu { path })?,
+    )
+    .context(error::FileParseJsonSnafu { path })
 }
 
-fn write_file<T>(path: &Path, json: &T) -> Result<()>
+async fn write_file<T>(path: &Path, json: &T) -> Result<()>
 where
     T: serde::Serialize,
 {
     // Use `tempfile::NamedTempFile::persist` to perform an atomic file write.
     let parent = path.parent().context(error::PathParentSnafu { path })?;
-    let mut writer =
+    let file =
         NamedTempFile::new_in(parent).context(error::FileTempCreateSnafu { path: parent })?;
-    serde_json::to_writer_pretty(&mut writer, json).context(error::FileWriteJsonSnafu { path })?;
-    writer
-        .write_all(b"\n")
+
+    let (file, tmp_path) = file.into_parts();
+    let mut file = tokio::fs::File::from_std(file);
+
+    let buf = serde_json::to_vec_pretty(json).context(error::FileWriteJsonSnafu { path })?;
+    file.write_all(&buf)
+        .await
         .context(error::FileWriteSnafu { path })?;
-    writer
+
+    let file = file.into_std().await;
+    NamedTempFile::from_parts(file, tmp_path)
         .persist(path)
         .context(error::FilePersistSnafu { path })?;
+
     Ok(())
 }
 
 // Walk the directory specified, building a map of filename to Target structs.
 // Hashing of the targets is done in parallel
-fn build_targets<P>(indir: P, follow_links: bool) -> Result<HashMap<TargetName, Target>>
+async fn build_targets<P>(indir: P, follow_links: bool) -> Result<HashMap<TargetName, Target>>
 where
     P: AsRef<Path>,
 {
-    let indir = indir.as_ref();
-    WalkDir::new(indir)
-        .follow_links(follow_links)
-        .into_iter()
-        .par_bridge()
-        .filter_map(|entry| match entry {
-            Ok(entry) => {
-                if entry.file_type().is_file() {
-                    Some(process_target(entry.path()))
-                } else {
-                    None
+    let indir = indir.as_ref().to_owned();
+
+    let (tx, rx) = tokio::sync::mpsc::channel(10);
```

Comment:
Is `10` the max parallelization? Should it be `num_cpu`?

```
    let (tx, rx) = tokio::sync::mpsc::channel(10);
```

Reply:
It's just the buffer depth of the producer. `WalkDir` will run on a special thread reserved for blocking tasks. I chose some value >1 to avoid starvation of the consumers while this thread isn't scheduled.
`tokio` with `rt-multi-thread` enabled will automatically scale the number of threads according to the number of cores. As the consuming futures are spawned as independent tasks, they will evenly distribute across those threads.